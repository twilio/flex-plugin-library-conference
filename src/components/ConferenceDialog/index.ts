import ConferenceDialog from './ConferenceDialog';
import { withTaskContext } from '@twilio/flex-ui';

export default withTaskContext(ConferenceDialog);
