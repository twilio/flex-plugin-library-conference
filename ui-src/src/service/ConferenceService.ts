/* Created for use with external transfer extensions

  merged from excellent work done by Terence Rogers
  https://github.com/trogers-twilio/plugin-external-conference-warm-transfer

*/
import ApiService from './ApiService';
import { EncodedParams } from '../types/Params';
import { FetchedCall, FetchedConferenceParticipant } from '../types/twilio-api';
import { ErrorManager, FlexPluginErrorType } from '../utils/ErrorManager';

export interface GetCallResponse {
  success: boolean;
  callProperties: FetchedCall;
}

export interface ParticipantResponse {
  success: boolean;
  callSid: string;
}

export interface RemoveParticipantResponse {
  success: boolean;
}

class ConferenceService extends ApiService {
  _toggleParticipantHold = async (conference: string, participantSid: string, hold: boolean): Promise<string> => {
    return new Promise((resolve, reject) => {
      const encodedParams: EncodedParams = {
        conference: encodeURIComponent(conference),
        participant: encodeURIComponent(participantSid),
        hold: encodeURIComponent(hold),
        Token: encodeURIComponent(this.manager.user.token),
      };

      this.fetchJsonWithReject<ParticipantResponse>(`${this.serverlessDomain}/conference/hold-participant`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: this.buildBody(encodedParams),
      })
        .then((response: ParticipantResponse) => {
          console.log(`${hold ? 'Hold' : 'Unhold'} successful for participant`, participantSid);
          resolve(response.callSid);
        })
        .catch((e) => {
          console.error(`Error ${hold ? 'holding' : 'unholding'} participant ${participantSid}\r\n`, e);
          ErrorManager.createAndProcessError(
            `Error ${hold ? 'holding' : 'unholding'} participant ${participantSid}\r\n`,
            {
              type: FlexPluginErrorType.serverless,
              description:
                e instanceof Error
                  ? `${e.message}`
                  : `Error ${hold ? 'holding' : 'unholding'} participant ${participantSid}\r\n`,
              context: 'Plugin.ConferenceService',
              wrappedError: e,
            },
          );
          reject(e);
        });
    });
  };

  setEndConferenceOnExit = async (
    conference: string,
    participantSid: string,
    endConferenceOnExit: boolean,
  ): Promise<string> => {
    return new Promise((resolve, reject) => {
      const encodedParams: EncodedParams = {
        conference: encodeURIComponent(conference),
        participant: encodeURIComponent(participantSid),
        endConferenceOnExit: encodeURIComponent(endConferenceOnExit),
        Token: encodeURIComponent(this.manager.user.token),
      };

      this.fetchJsonWithReject<ParticipantResponse>(`${this.serverlessDomain}/conference/update-participant`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: this.buildBody(encodedParams),
      })
        .then((response: ParticipantResponse) => {
          console.log(`Participant ${participantSid} updated:\r\n`, response);
          resolve(response.callSid);
        })
        .catch((e) => {
          console.error(`Error updating participant ${participantSid}\r\n`, e);
          ErrorManager.createAndProcessError(`Error updating participant ${participantSid}\r\n`, {
            type: FlexPluginErrorType.serverless,
            description: e instanceof Error ? `${e.message}` : `Error updating participant ${participantSid}\r\n`,
            context: 'Plugin.ConferenceService',
            wrappedError: e,
          });
          reject(e);
        });
    });
  };

  addParticipant = async (taskSid: string, from: string, to: string): Promise<string> => {
    return new Promise((resolve, reject) => {
      const encodedParams: EncodedParams = {
        taskSid: encodeURIComponent(taskSid),
        from: encodeURIComponent(from),
        to: encodeURIComponent(to),
        Token: encodeURIComponent(this.manager.user.token),
      };

      this.fetchJsonWithReject<ParticipantResponse>(`${this.serverlessDomain}/conference/add-participant`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: this.buildBody(encodedParams),
      })
        .then((response: ParticipantResponse) => {
          console.log('Participant added:\r\n  ', response);
          resolve(response.callSid);
        })
        .catch((e) => {
          console.log('There is an error while adding participant', e);
          ErrorManager.createAndProcessError(`Error while adding participant ${to}`, {
            type: FlexPluginErrorType.serverless,
            description: e instanceof Error ? `${e.message}` : `Error while adding participant ${to}`,
            context: 'Plugin.ConferenceService',
            wrappedError: e,
          });
          reject(e);
        });
    });
  };

  holdParticipant = (conference: string, participantSid: string): Promise<string> => {
    return this._toggleParticipantHold(conference, participantSid, true);
  };

  unholdParticipant = (conference: string, participantSid: string): Promise<string> => {
    return this._toggleParticipantHold(conference, participantSid, false);
  };

  removeParticipant = (conference: string, participantSid: string): Promise<string> => {
    return new Promise((resolve, reject) => {
      const encodedParams: EncodedParams = {
        conference: encodeURIComponent(conference),
        participant: encodeURIComponent(participantSid),
        Token: encodeURIComponent(this.manager.user.token),
      };

      this.fetchJsonWithReject<RemoveParticipantResponse>(`${this.serverlessDomain}/conference/remove-participant`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: this.buildBody(encodedParams),
      })
        .then((response: RemoveParticipantResponse) => {
          console.log(`Participant ${participantSid} removed from conference`);
          resolve(participantSid);
        })
        .catch((e) => {
          console.error(`Error removing participant ${participantSid} from conference\r\n`, e);
          ErrorManager.createAndProcessError(`Error removing participant ${participantSid} from conference\r\n`, {
            type: FlexPluginErrorType.serverless,
            description:
              e instanceof Error ? `${e.message}` : `Error removing participant ${participantSid} from conference\r\n`,
            context: 'Plugin.ConferenceService',
            wrappedError: e,
          });
          reject(e);
        });
    });
  };

  getCallProperties = async (callSid: string): Promise<FetchedCall> => {
    return new Promise((resolve, reject) => {
      const encodedParams: EncodedParams = {
        callSid: encodeURIComponent(callSid),
        Token: encodeURIComponent(this.manager.user.token),
      };

      this.fetchJsonWithReject<GetCallResponse>(`${this.serverlessDomain}/conference/get-call-properties`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: this.buildBody(encodedParams),
      })
        .then((resp: GetCallResponse) => {
          resolve(resp.callProperties);
        })
        .catch((e) => {
          console.log('There is an error', e);
          ErrorManager.createAndProcessError(`Error fetching call properties for call ${callSid}\r\n`, {
            type: FlexPluginErrorType.serverless,
            description: e instanceof Error ? `${e.message}` : `Error fetching call properties for call ${callSid}\r\n`,
            context: 'Plugin.ConferenceService',
            wrappedError: e,
          });
          reject(e);
        });
    });
  };
}

export default new ConferenceService();
