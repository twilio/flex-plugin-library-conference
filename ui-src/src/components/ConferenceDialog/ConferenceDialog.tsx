import * as React from 'react';
import { Actions, Manager, ITask, useFlexSelector } from '@twilio/flex-ui';
import { useDispatch } from 'react-redux';
import { AppState } from '../../flex-hooks/states/ConferenceSlice';
import ConferenceService from '../../service/ConferenceService';

import { useUID } from '@twilio-paste/core/uid-library';
import { Box } from '@twilio-paste/core/box';
import { Button } from '@twilio-paste/core/button';
import { Input } from '@twilio-paste/core/input';
import { Label } from '@twilio-paste/core/label';
import { Modal, ModalBody, ModalFooter, ModalFooterActions, ModalHeader, ModalHeading } from '@twilio-paste/core/modal';
import { PopoverContainer, Popover, PopoverButton } from '@twilio-paste/core/popover';
import { InformationIcon } from '@twilio-paste/icons/cjs/InformationIcon';
import { Text } from '@twilio-paste/core/text';
import { addConnectingParticipant } from '../../flex-hooks/states/ConferenceSlice';
import { ErrorManager, FlexPluginErrorType } from '../../utils/ErrorManager';

export interface OwnProps {
  task?: ITask;
}

const ConferenceDialog = (props: OwnProps) => {
  const [conferenceTo, setConferenceTo] = React.useState('');

  const componentViewStates = useFlexSelector((state: AppState) => state.flex.view.componentViewStates);
  const phoneNumber = useFlexSelector((state: AppState) => state.flex.worker.attributes.phone);

  const conferenceDialogState = componentViewStates && componentViewStates.ConferenceDialog;
  const isOpen = (conferenceDialogState && conferenceDialogState.isOpen) || false;

  const dispatch = useDispatch();
  const modalHeadingID = useUID();
  const inputRef = React.createRef<HTMLInputElement>();
  const inputID = useUID();

  const handleClose = () => {
    closeDialog();
  };

  const handleButtonClose = (e: React.MouseEvent<HTMLElement>) => {
    closeDialog();
    if (e) e.preventDefault();
  };

  const closeDialog = () => {
    Actions.invokeAction('SetComponentState', {
      name: 'ConferenceDialog',
      state: { isOpen: false },
    });
  };

  const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const key = e.key;

    if (key === 'Enter') {
      addConferenceParticipant();
      closeDialog();
      e.preventDefault();
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setConferenceTo(value);
  };

  const handleDialButton = (e: React.MouseEvent<HTMLElement>) => {
    addConferenceParticipant();
    closeDialog();
    e.preventDefault();
  };

  const addConferenceParticipant = async () => {
    const { task } = props;

    if (!task) return;

    let mainConferenceSid = task.attributes.conference ? task.attributes.conference.sid : null;

    if (!mainConferenceSid && task.conference) {
      mainConferenceSid = task.conference.conferenceSid;
    }

    let from;
    if (phoneNumber) {
      from = phoneNumber;
    } else {
      from = Manager.getInstance().serviceConfiguration.outbound_call_flows.default.caller_id;
    }

    // Adding entered number to the conference
    console.log(`Adding ${conferenceTo} to conference`);
    let participantCallSid;
    try {
      participantCallSid = await ConferenceService.addParticipant(mainConferenceSid, from, conferenceTo);
      dispatch(
        addConnectingParticipant({
          callSid: participantCallSid,
          conferenceSid: mainConferenceSid,
          phoneNumber: conferenceTo,
        }),
      );
    } catch (e) {
      console.error('Error adding conference participant:', e);
      ErrorManager.createAndProcessError('Could not add participant to conference', {
        type: FlexPluginErrorType.serverless,
        description: e instanceof Error ? `${e.message}` : 'Could not add participant to conference',
        context: 'Plugin.ConferenceDialog',
        wrappedError: e,
      });
    }

    setConferenceTo('');
  };

  return (
    <Modal
      ariaLabelledby={modalHeadingID}
      isOpen={isOpen}
      onDismiss={handleClose}
      // set initial focus here
      initialFocusRef={inputRef}
      size="default"
      data-testid="modal-header"
    >
      <ModalHeader>
        <ModalHeading as="h3" id={modalHeadingID}>
          Add Conference Participant
        </ModalHeading>
      </ModalHeader>
      <ModalBody>
        <Box as="form">
          <Box display="flex" alignItems="center" marginBottom="space40">
            <Label htmlFor={inputID} marginBottom="space0">Phone Number</Label>
            <PopoverContainer baseId={`hint-popover`} placement="right">
              <PopoverButton variant="secondary_icon" size="icon_small">
                <InformationIcon decorative={false} title="Help link" color="colorTextLink" />
              </PopoverButton>
              <Popover aria-label="Popover" element="plugin-hint-popover">
                <Text as="span" marginRight="space20">
                  Phone number with country code: +1xxxxxxxxxx
                </Text>
              </Popover>
            </PopoverContainer>
          </Box>
          <Input
            id={inputID}
            value={conferenceTo}
            data-testid="phoneNumberInput"
            // assign the target ref here
            ref={inputRef}
            onChange={handleChange}
            onKeyPress={handleKeyPress}
            type="text"
          />
        </Box>
      </ModalBody>
      <ModalFooter>
        <ModalFooterActions>
          <Button variant="secondary" onClick={handleButtonClose} title="Cancel">
            Cancel
          </Button>
          <Button variant="primary" onClick={handleDialButton} title="Dial">
            Dial
          </Button>
        </ModalFooterActions>
      </ModalFooter>
    </Modal>
  );
};

export default ConferenceDialog;
