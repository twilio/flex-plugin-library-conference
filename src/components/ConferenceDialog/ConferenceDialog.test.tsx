import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ITask, Actions } from '@twilio/flex-ui';
import ConferenceDialog from './ConferenceDialog';
import ConferenceService from '../../service/ConferenceService';
import { useSelector, useDispatch } from 'react-redux';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch
}));
jest.mock('@twilio/flex-ui', () => {
    return {
      __esModule: true,
      Actions:{
        addListener: jest.fn(),
        removeListener: jest.fn(),
        invokeAction: jest.fn(),
      },
      Manager:{
        getInstance:jest.fn()
      },
      useFlexSelector: jest.fn(() => {return {ConferenceDialog: {isOpen: true}}})
    };
  });


describe('Conference Dialog', () => {
  // const t = { sid: '1672673' } as unknown as ITask;
  it('should render correct snapshot', () => {
    const wrapper = render(<ConferenceDialog />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should click Cancel button', () => {
    const invokeAction = jest.spyOn(Actions, 'invokeAction');
    const { getByTitle } = render(<ConferenceDialog />);
    const cancelButton = getByTitle('Cancel').closest('button');
    if (cancelButton) {
      fireEvent.click(cancelButton);
    }
    expect(invokeAction).toHaveBeenCalledWith('SetComponentState',{
      name: 'ConferenceDialog',
      state: { isOpen: false }
    });
  });

  it('should click Cancel button', () => {
    const invokeAction = jest.spyOn(Actions, 'invokeAction');
    const { getByTitle } = render(<ConferenceDialog />);
    const cancelButton = getByTitle('Cancel').closest('button');
    if (cancelButton) {
      fireEvent.click(cancelButton);
    }
    expect(invokeAction).toHaveBeenCalledWith('SetComponentState',{
      name: 'ConferenceDialog',
      state: { isOpen: false }
    });
  });

  it('should close dialog on esc press', () => {
    const invokeAction = jest.spyOn(Actions, 'invokeAction');
    const { getByTestId } = render(<ConferenceDialog />);
    fireEvent.click(getByTestId('modal-header').querySelector('button') as HTMLButtonElement);
    expect(invokeAction).toHaveBeenCalledWith('SetComponentState',{
      name: 'ConferenceDialog',
      state: { isOpen: false }
    });
  });

})