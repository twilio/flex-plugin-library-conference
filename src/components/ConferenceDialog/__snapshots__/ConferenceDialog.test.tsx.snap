// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conference Dialog should render correct snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-hidden="true"
    />
    <reach-portal>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div>
          <div
            class="css-159lja1"
            data-paste-element="MODAL_OVERLAY"
            data-reach-dialog-overlay=""
            style="opacity: 0;"
            variant="default"
          >
            <div
              aria-labelledby="1"
              aria-modal="true"
              class="css-trckgm"
              data-paste-element="MODAL"
              data-reach-dialog-content=""
              data-testid="modal-header"
              role="dialog"
              style="opacity: 0; transform: scale(0.675);"
              tabindex="-1"
            >
              <div
                class="css-1b9e17t"
                data-paste-element="MODAL_HEADER"
              >
                <div
                  class="css-1ibkied"
                  data-paste-element="FLEX"
                >
                  <div
                    class="css-1072z4j"
                    data-paste-element="FLEX"
                  >
                    <h3
                      class="css-n8ur01"
                      data-paste-element="MODAL_HEADING"
                      id="1"
                    >
                      Add Conference Participant
                    </h3>
                  </div>
                  <button
                    aria-busy="false"
                    class="css-48drkk"
                    data-paste-element="MODAL_HEADER_CLOSE_BUTTON"
                    type="button"
                  >
                    <span
                      class="css-z6jp9v"
                      data-paste-element="BOX"
                    >
                      <span
                        class="css-yrazta"
                        data-paste-element="MODAL_HEADER_CLOSE_ICON"
                      >
                        <svg
                          aria-hidden="true"
                          aria-labelledby="CloseIcon-3"
                          height="100%"
                          role="img"
                          viewBox="0 0 20 20"
                          width="100%"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M15.16 13.514L11.645 10l3.515-3.514a1.165 1.165 0 00-1.646-1.646L10 8.355 6.486 4.84A1.165 1.165 0 004.84 6.486L8.355 10 4.84 13.514a1.165 1.165 0 001.646 1.646L10 11.645l3.514 3.515a1.165 1.165 0 001.646-1.646z"
                            fill="currentColor"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                      <span
                        class="css-1t35qh6"
                        data-paste-element="BOX"
                      >
                        Close modal
                      </span>
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="css-137sttb"
                data-paste-element="MODAL_BODY"
              >
                <form
                  class="css-roynbj"
                  data-paste-element="BOX"
                >
                  <label
                    class="css-rwo46e"
                    data-paste-element="LABEL"
                    for="2"
                  >
                    <span
                      class="css-1sekzvp"
                      data-paste-element="MEDIA_OBJECT"
                    >
                      <span
                        class="css-1cywlpt"
                        data-paste-element="MEDIA_BODY"
                      >
                        Phone Number
                      </span>
                    </span>
                  </label>
                  <div
                    class="css-nrrnrg"
                    data-paste-element="INPUT"
                  >
                    <input
                      class="css-9se8rz"
                      data-paste-element="INPUT_ELEMENT"
                      id="2"
                      type="text"
                      value=""
                    />
                  </div>
                </form>
              </div>
              <div
                class="css-1aspgdz"
                data-paste-element="MODAL_FOOTER"
              >
                <div
                  class="css-rt1g17"
                  data-paste-element="MODAL_FOOTER_ACTIONS"
                >
                  <div
                    class="css-10brdie"
                    data-paste-element="MODAL_FOOTER_ACTIONS_ITEM"
                  >
                    <button
                      aria-busy="false"
                      class="css-kzhkm0"
                      data-paste-element="BUTTON"
                      title="Cancel"
                      type="button"
                    >
                      <span
                        class="css-z6jp9v"
                        data-paste-element="BOX"
                      >
                        Cancel
                      </span>
                    </button>
                  </div>
                  <div
                    class="css-roynbj"
                    data-paste-element="MODAL_FOOTER_ACTIONS_ITEM"
                  >
                    <button
                      aria-busy="false"
                      class="css-175kaqp"
                      data-paste-element="BUTTON"
                      title="Dial"
                      type="button"
                    >
                      <span
                        class="css-z6jp9v"
                        data-paste-element="BOX"
                      >
                        Dial
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </reach-portal>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
